atom/index.d.ts

```diff
@@ -99,6 +99,64 @@ export interface WritableAtom<Value = any> extends ReadableAtom<Value> {
 
 export type Atom<Value = any> = ReadableAtom<Value> | WritableAtom<Value>
 
+export interface Autosubscribe<Value = any> {
+  /**
+   * @return The computed store's value
+   */
+  (): Value
+  /**
+   * Autosubscribes the associated computed store to the given atom.
+   *
+   * @param atom
+   */
+  <V>(atom: ReadableAtom<V>): V
+  /**
+   * Runs fn when the computed store is unmounted or when the computed store's cb is next run
+   *
+   * When .stale() is true, saving fn callback is deactivated.
+   *
+   * @param fn
+   */
+  off(fn: () => any): Autosubscribe<Value>
+  /**
+   * Immediately runs fn & calls fn when the computed store is mounted.
```

Comment:
We already have event system with `onMount`, `onStart` and `onStop`.

Do you avoid `onMount` for async support?

Reply:
I did not know about `onStart` & `onStop`.

The issue with `onMount` is that it's passed the computed atom as an argument, which is not reliably accessible within the computed callback.

The use case is to be able to use `onMount`, `onStart`, & `onStop` from within the computed callback, so the computed's value can be directly set within the `onMount`. I added `on` & `off` not knowing about `onStart` & `onStop`. Perhaps `on` can be renamed to `onStart` & `off` can be renamed to `onStop`.

Reply:
On `Autosubscribe`, I renamed the methods `.on` to `.onStart` & `.off` to `.onStop`, keeping the names already used in the event system.

Reply:
The problem is not in the name, but in duplication. Now we have the methods to do the same.

Reply:
Right. There are a few problems I ran into.

The existing `onStart` & `onStop` cannot be used inside of the computed callback:

- The `$computed` object is not accessible inside of the callback, which `onStart` & `onStop` requires
- The existing `onStart` & `onStop` have no stale checks so new `onStart` & `onStop` listeners would be bound every time the callback runs

This means the usage requires the `$computed` to be created first. Since `$computed` is a `ReadableAtom`, there is no way in the type system to directly set `$computed.value` or call `$computed.set`...In js, these props are there but a computed is not meant to be writable.

This means an `$atom` must be created which the `$computed` listens to.

Here is a comparison of the two approaches:

Using the existing `onStart` & `onStop`.

```ts
import { atom, computed, onStart, onStop } from 'nanostores'
let $colorSchemeAtom = atom()
let watch = window.matchMedia('(prefers-color-scheme: dark)')
let onchange = evt => $colorSchemeAtom(evt.matches ? 'dark' : 'light')
let $colorScheme = computed(use => use($atom))
onchange(watch)
onStart($computed, () => watch.addEventListener('change', onchange))
onStop($computed, () => watch.removeEventListener('change', onchange))
```

Using the autosubscribe.onStart & autosubscribe.onStop

```ts
import { computed } from 'nanostores'
let $colorScheme = computed(use => {
 let watch = window.matchMedia('(prefers-color-scheme: dark)')
 let onchange = evt =>use.save(evt.matches ? 'dark' : 'light')
 use.onStart(() => watch.addEventListener('change', onchange))
 use.onStop(() => watch.removeEventListener('change', onchange))
 return onchange(watch)
})
```

Having `.onStart` & `.onStop` in autosubscribe allows the `computed` callback to be completely self-contained without requiring the extra `$colorSchemeAtom` to be created & written to.

Reply:
An interesting impact of `.onStart`/`.onStop` inside of the computed callback is it enables a `$computed` to be driven without any dependencies on an `$atom`, bridging to non-nanostores event systems.

Reply:
I also don’t like that current PR is huge and contain many different API changes :(. It is impossible to review and accept it in the current huge form.

Reply:
> I don’t agree that it is more convenient.

I'm coming from an object pov with regards to less lines of code and self-containment vs multiple variables in the calling scope. The two example effectively do exactly the same thing from a declarative pov. There is a difference that I'll explain further down.

> computed() stores were created as derivatives stores from other stores. You are using them for any “smart” stores which calc values according to system API. I think it is better to separated derivatives, simple stores, and smart stores.

I don't see a need to create a distinction between derivative & smart stores. For example, what if you want both derivative & smart store behavior? Which one would be used? To me, I'm taking the term "computed" at face value. It is a computationally created atom. The computation can include autosubscribing derivations from other stores and/or listen to system events. Sure, there is historical usage, coming from Svelte I believe, of the derived store being derived from other stores. I think you made a brilliant choice to use "computed" instead. This PR just extends `computed` to a fuller meaning of the term "computed".

> computed() callback is calling on every parent-store changes.

Yes that is true. In this PR, if there are no parent stores, then the callback is called once.

> In your API store#onStart works like a React’s hook (works only first time, then ignored).

I'm not able to reconcile your interpretation with mine. We can further discuss to align our perspectives. The API is "re-entrant", meaning there is only binding of `.onStart`/`.onStop` at a time. Let's say the `computed` is both derived from another `$store` & uses `.onStart`/`.onStop`. When the parent `$store` updates, the callback is run again, which means `.onStart`/`.onStop` is called again. This api is re-entrant, so the previous binding of `.onStart`/`.onStop` is removed & then readded when the callback is invoked.

> It is not very convenient API.

It should effectively act the same as the existing API, except it allows `computed` to be self-contained without needing extra imperative logic & objects such as the `$colorSchemeAtom` along with the variables in the calling scope. Instead, the logic is self-contained within the `computed` callback, meaning the calling scope is not polluted with these imperative variables that only apply to the `$colorScheme`.

> I also don’t like that current PR is huge and contain many different API changes :(. It is impossible to review and accept it in the current huge form.

I hope you can reconsider. This PR does not have any breaking changes & is additive so existing usages of nanostores should not have any issues. I admit that this is a large change, but it's mainly to flesh out unsupported use cases such as autosubscription & async computeds in a cohesive way.

I bundled the `async` behavior along with the autosubscribe behavior to inform the design of the system. I & several others have bespoke functions to add async behavior to `computed`, so I think it is a relevant use case. If async `computeds` were available in nanostores, it will be more widely used because it's a powerful addition. Adding `async` behavior introduces the notions of staleness & re-entrancy, so this PR works through those issues.

Some convenience methods were added to `autosubscribe`:

- `()`: the current value of the `$computed` for deep-equal & diff checking
- `.set`/`.save` to support intermediate async states such as loading
- `.stale` to prevent old callback invocations to clobber the value of newer callback invocations
- `.onStart`/`.onStop` to support self-contained computeds using system apis

Having used this in my code-bases, I found all of these to be useful. `.onStart`/`.onStop` is not strictly necessary, but it does cut down on lines of code, & is self-contained meaning the functional pieces are all bundled within the callback. It also faithfully implements the term `computed` (that was a brilliant rename on your part) as a store that is computationally defined.